File: sentinelnet/src/services/signingProxy.ts
import axios, { AxiosError, AxiosInstance } from 'axios';
import https from 'https';
import fs from 'fs';
import crypto from 'crypto';
import logger from '../logger';

const LOCAL_SIGNER_PREFIX = 'local-ed25519:';
const DEFAULT_TIMEOUT_MS = 3000;

type KmsSignResponse = { signature_b64?: string; signer_id?: string };
type KmsVerifyResponse = { verified?: boolean };

type LocalKeyPair = {
  privateKey: crypto.KeyObject;
  publicKey: crypto.KeyObject;
  publicBytes: Buffer;
};

export class SigningProxy {
  private readonly endpoint: string | undefined;
  private readonly keyB64?: string;
  private readonly client: AxiosInstance;
  private localKey?: LocalKeyPair;

  constructor() {
    this.endpoint = (process.env.SENTINEL_KMS_ENDPOINT || '').replace(/\/+$/, '') || undefined;
    this.keyB64 = process.env.SENTINEL_SIGNER_KEY_B64;

    const timeoutMs =
      parseInt(process.env.KMS_TIMEOUT_MS || '', 10) > 0
        ? parseInt(process.env.KMS_TIMEOUT_MS || '', 10)
        : DEFAULT_TIMEOUT_MS;

    const httpsAgent = buildMtlsAgent();
    this.client = axios.create({
      baseURL: this.endpoint,
      timeout: timeoutMs,
      httpsAgent,
      validateStatus: (status) => status >= 200 && status < 300,
    });
  }

  async sign(payload: Buffer): Promise<{ signatureB64: string; signerId: string }> {
    if (!payload || payload.length === 0) {
      throw new Error('payload is required for signing');
    }

    if (this.endpoint) {
      try {
        return await this.signWithKms(payload);
      } catch (err) {
        logger.warn('KMS signing failed, attempting fallback', { error: (err as Error).message });
      }
    }

    return this.signLocal(payload);
  }

  async verify(payload: Buffer, signatureB64: string, signerId: string): Promise<void> {
    if (signerId.startsWith(LOCAL_SIGNER_PREFIX)) {
      return this.verifyLocal(payload, signatureB64);
    }
    if (!this.endpoint) {
      throw new Error('KMS endpoint not configured; cannot verify remote signature');
    }
    return this.verifyWithKms(payload, signatureB64, signerId);
  }

  private async signWithKms(payload: Buffer): Promise<{ signatureB64: string; signerId: string }> {
    const body = { payload_b64: payload.toString('base64') };
    const resp = await this.postWithRetry('/sign', body);
    const data = (resp?.data || {}) as KmsSignResponse;
    if (!data.signature_b64 || !data.signer_id) {
      throw new Error('kms response missing signature_b64 or signer_id');
    }
    return { signatureB64: data.signature_b64, signerId: data.signer_id };
  }

  private async verifyWithKms(
    payload: Buffer,
    signatureB64: string,
    signerId: string,
  ): Promise<void> {
    const body = {
      payload_b64: payload.toString('base64'),
      signature_b64: signatureB64,
      signer_id: signerId,
    };
    const resp = await this.postWithRetry('/verify', body);
    const data = (resp?.data || {}) as KmsVerifyResponse;
    if (!data.verified) {
      throw new Error('kms verification failed');
    }
  }

  private async signLocal(payload: Buffer): Promise<{ signatureB64: string; signerId: string }> {
    const key = this.ensureLocalKey();
    const signature = crypto.sign(null, payload, key.privateKey);
    const signerId = `${LOCAL_SIGNER_PREFIX}${shortSha(key.publicBytes)}`;
    return { signatureB64: signature.toString('base64'), signerId };
  }

  private async verifyLocal(payload: Buffer, signatureB64: string): Promise<void> {
    const key = this.ensureLocalKey();
    const sig = Buffer.from(signatureB64, 'base64');
    const ok = crypto.verify(null, payload, key.publicKey, sig);
    if (!ok) {
      throw new Error('local signature verification failed');
    }
  }

  private async postWithRetry(path: string, body: any) {
    let lastErr: any;
    let delay = 100;
    for (let attempt = 0; attempt < 2; attempt += 1) {
      if (attempt > 0) {
        await new Promise((resolve) => setTimeout(resolve, delay));
        delay *= 2;
      }
      try {
        return await this.client.post(path, body);
      } catch (err) {
        lastErr = err;
        if (!this.shouldRetry(err) || attempt === 1) {
          throw err;
        }
      }
    }
    throw lastErr;
  }

  private shouldRetry(err: any): boolean {
    const axiosErr = err as AxiosError;
    const status = axiosErr.response?.status || 0;
    if (status >= 500) return true;
    const code = (axiosErr as any)?.code;
    const retryable = ['ECONNRESET', 'ETIMEDOUT', 'EAI_AGAIN', 'ENETUNREACH'];
    if (code && retryable.includes(code)) return true;
    if (axiosErr.response == null && axiosErr.request) return true;
    return false;
  }

  private ensureLocalKey(): LocalKeyPair {
    if (this.localKey) {
      return this.localKey;
    }
    if (!this.keyB64) {
      throw new Error('fallback key not configured (SENTINEL_SIGNER_KEY_B64)');
    }
    const raw = Buffer.from(this.keyB64.trim(), 'base64');
    if (![32, 64].includes(raw.length)) {
      throw new Error(`unexpected ed25519 key length ${raw.length}`);
    }
    const seed = raw.slice(0, 32);
    const privateKey = crypto.createPrivateKey({
      key: buildEd25519PKCS8(seed),
      format: 'der',
      type: 'pkcs8',
    });
    let publicBytes: Buffer;
    let publicKey: crypto.KeyObject;
    if (raw.length === 64) {
      publicBytes = raw.slice(32);
      publicKey = crypto.createPublicKey({
        key: buildEd25519SPKI(publicBytes),
        format: 'der',
        type: 'spki',
      });
    } else {
      publicKey = crypto.createPublicKey(privateKey);
      publicBytes = extractPublicKeyBytes(publicKey);
    }
    this.localKey = { privateKey, publicKey, publicBytes };
    return this.localKey;
  }
}

function buildMtlsAgent(): https.Agent | undefined {
  const certEnv = process.env.SENTINEL_CLIENT_CERT;
  const keyEnv = process.env.SENTINEL_CLIENT_KEY;
  const caEnv = process.env.SENTINEL_CA_CERT;

  if (!certEnv || !keyEnv) {
    if (process.env.DEV_SKIP_MTLS === 'true') {
      logger.info('DEV_SKIP_MTLS enabled; not configuring mTLS for KMS client');
    }
    return undefined;
  }

  try {
    const cert = readValueOrFile(certEnv);
    const key = readValueOrFile(keyEnv);
    const ca = caEnv ? readValueOrFile(caEnv) : undefined;
    return new https.Agent({
      cert,
      key,
      ca,
      minVersion: 'TLSv1.2',
      rejectUnauthorized: true,
    });
  } catch (err) {
    logger.warn('failed to configure mTLS agent; proceeding without client cert', {
      error: (err as Error).message,
    });
    return undefined;
  }
}

function readValueOrFile(value: string): Buffer {
  if (fs.existsSync(value)) {
    return fs.readFileSync(value);
  }
  if (value.includes('BEGIN')) {
    return Buffer.from(value);
  }
  try {
    const decoded = Buffer.from(value, 'base64');
    if (decoded.length > 0) {
      return decoded;
    }
  } catch {
    // ignore and return raw
  }
  return Buffer.from(value);
}

function buildEd25519PKCS8(seed: Buffer): Buffer {
  const prefix = Buffer.from('302e020100300506032b657004220420', 'hex');
  return Buffer.concat([prefix, seed]);
}

function buildEd25519SPKI(publicKey: Buffer): Buffer {
  const prefix = Buffer.from('302a300506032b6570032100', 'hex');
  return Buffer.concat([prefix, publicKey]);
}

function extractPublicKeyBytes(pub: crypto.KeyObject): Buffer {
  const der = pub.export({ format: 'der', type: 'spki' }) as Buffer;
  return der.slice(-32);
}

function shortSha(buf: Buffer): string {
  const hash = crypto.createHash('sha256').update(buf).digest('hex');
  return hash.slice(0, 8);
}

export default SigningProxy;

File: kernel/internal/signing/signing_proxy.go
package signing

import (
	"bytes"
	"context"
	"crypto/ed25519"
	"crypto/sha256"
	"crypto/tls"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

// SigningProxy signs payloads by delegating to a KMS HTTP endpoint with an mTLS-capable client.
// If the endpoint is not configured or fails, it falls back to a local Ed25519 key provided via env.
type SigningProxy struct {
	client   *http.Client
	endpoint string
	keyB64   string
	keyID    string
}

const (
	defaultKMSTimeout = 3 * time.Second
	localSignerPrefix = "local-ed25519:"
)

// kmsSignResponse is the expected KMS /sign response contract.
type kmsSignResponse struct {
	SignatureB64 string `json:"signature_b64"`
	SignerID     string `json:"signer_id"`
}

// kmsVerifyResponse matches the optional KMS /verify contract.
type kmsVerifyResponse struct {
	Verified bool `json:"verified"`
}

// NewSigningProxyFromEnv builds a SigningProxy using environment-driven configuration.
// Environment variables:
//   - KERNEL_KMS_ENDPOINT: base URL of the KMS service (no trailing slash required)
//   - KERNEL_KMS_KEY_ID: optional logical key identifier passed through to KMS
//   - KERNEL_SIGNER_KEY_B64: base64-encoded Ed25519 private key (fallback dev path)
//   - KERNEL_CLIENT_CERT / KERNEL_CLIENT_KEY / KERNEL_CA_CERT: PEM contents or file paths for mTLS
//   - KMS_TIMEOUT_MS: optional request timeout in milliseconds (default 3000ms)
func NewSigningProxyFromEnv() (*SigningProxy, error) {
	endpoint := strings.TrimRight(os.Getenv("KERNEL_KMS_ENDPOINT"), "/")
	keyID := os.Getenv("KERNEL_KMS_KEY_ID")
	keyB64 := os.Getenv("KERNEL_SIGNER_KEY_B64")
	timeout := defaultKMSTimeout
	if v := os.Getenv("KMS_TIMEOUT_MS"); v != "" {
		if ms, err := strconv.Atoi(v); err == nil && ms > 0 {
			timeout = time.Duration(ms) * time.Millisecond
		} else {
			log.Printf("[signing] invalid KMS_TIMEOUT_MS=%q; using default %s", v, defaultKMSTimeout)
		}
	}

	certEnv := os.Getenv("KERNEL_CLIENT_CERT")
	keyEnv := os.Getenv("KERNEL_CLIENT_KEY")
	caEnv := os.Getenv("KERNEL_CA_CERT")

	client, err := buildHTTPClient(certEnv, keyEnv, caEnv, timeout)
	if err != nil {
		return nil, err
	}

	return &SigningProxy{
		client:   client,
		endpoint: endpoint,
		keyB64:   keyB64,
		keyID:    keyID,
	}, nil
}

// Sign signs the provided payload. It prefers KMS; on failure (or when no endpoint is configured)
// it falls back to the local Ed25519 key if available.
func (s *SigningProxy) Sign(payload []byte) (string, string, error) {
	if len(payload) == 0 {
		return "", "", errors.New("signing payload is empty")
	}

	// Attempt KMS first if configured.
	if s.endpoint != "" {
		if sig, signer, err := s.signWithKMS(payload); err == nil {
			return sig, signer, nil
		} else {
			log.Printf("[signing] KMS signing failed: %v; attempting fallback", err)
		}
	}

	// Fallback to local Ed25519
	return s.signWithLocal(payload)
}

// Verify validates the provided signature for the payload. If the signerId indicates a local
// signature, verification is performed locally. Otherwise a KMS /verify call is attempted.
func (s *SigningProxy) Verify(payload []byte, signatureB64 string, signerID string) error {
	if strings.HasPrefix(signerID, localSignerPrefix) {
		return s.verifyLocal(payload, signatureB64)
	}

	if s.endpoint == "" {
		return errors.New("KMS endpoint not configured; cannot verify remote signature")
	}

	return s.verifyWithKMS(payload, signatureB64, signerID)
}

func (s *SigningProxy) signWithKMS(payload []byte) (string, string, error) {
	if s.client == nil {
		return "", "", errors.New("http client not configured")
	}
	url := s.endpoint + "/sign"

	reqBody := map[string]string{
		"payload_b64": base64.StdEncoding.EncodeToString(payload),
	}
	if s.keyID != "" {
		reqBody["key_id"] = s.keyID
	}

	var resp kmsSignResponse
	if err := s.postWithRetry(url, reqBody, &resp); err != nil {
		return "", "", err
	}

	if resp.SignatureB64 == "" || resp.SignerID == "" {
		return "", "", errors.New("kms response missing signature or signer_id")
	}

	return resp.SignatureB64, resp.SignerID, nil
}

func (s *SigningProxy) verifyWithKMS(payload []byte, signatureB64 string, signerID string) error {
	if s.client == nil {
		return errors.New("http client not configured")
	}
	url := s.endpoint + "/verify"

	reqBody := map[string]string{
		"payload_b64":   base64.StdEncoding.EncodeToString(payload),
		"signature_b64": signatureB64,
		"signer_id":     signerID,
	}

	var resp kmsVerifyResponse
	if err := s.postWithRetry(url, reqBody, &resp); err != nil {
		return err
	}
	if !resp.Verified {
		return errors.New("kms verification failed")
	}
	return nil
}

func (s *SigningProxy) signWithLocal(payload []byte) (string, string, error) {
	priv, pub, err := decodeEd25519Key(s.keyB64)
	if err != nil {
		return "", "", fmt.Errorf("ed25519 fallback unavailable: %w", err)
	}
	sig := ed25519.Sign(priv, payload)
	signerID := fmt.Sprintf("%s%x", localSignerPrefix, shortSHA(pub))
	return base64.StdEncoding.EncodeToString(sig), signerID, nil
}

func (s *SigningProxy) verifyLocal(payload []byte, signatureB64 string) error {
	pub, err := derivePublicKey(s.keyB64)
	if err != nil {
		return fmt.Errorf("ed25519 verify failed: %w", err)
	}
	sig, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil {
		return fmt.Errorf("invalid base64 signature: %w", err)
	}
	if !ed25519.Verify(pub, payload, sig) {
		return errors.New("signature verification failed")
	}
	return nil
}

// postWithRetry performs a POST with a single retry on transient errors or 5xx responses.
func (s *SigningProxy) postWithRetry(url string, body interface{}, out interface{}) error {
	var lastErr error
	backoff := 100 * time.Millisecond
	for attempt := 0; attempt < 2; attempt++ {
		if attempt > 0 {
			time.Sleep(backoff)
			backoff *= 2
		}
		if err := s.postJSON(url, body, out); err != nil {
			lastErr = err
			var netErr net.Error
			if errors.As(err, &netErr) && (netErr.Timeout() || netErr.Temporary()) {
				continue
			}
			var httpErr *httpStatusError
			if errors.As(err, &httpErr) && httpErr.ShouldRetry() {
				continue
			}
			return err
		}
		return nil
	}
	return lastErr
}

func (s *SigningProxy) postJSON(url string, body interface{}, out interface{}) error {
	buf := &bytes.Buffer{}
	if err := json.NewEncoder(buf).Encode(body); err != nil {
		return err
	}

	ctx, cancel := context.WithTimeout(context.Background(), s.client.Timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, buf)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 500 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return &httpStatusError{StatusCode: resp.StatusCode, Body: string(b)}
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return fmt.Errorf("kms returned http %d: %s", resp.StatusCode, string(b))
	}

	if out != nil {
		if err := json.NewDecoder(resp.Body).Decode(out); err != nil {
			return fmt.Errorf("kms decode error: %w", err)
		}
	}
	return nil
}

// buildHTTPClient constructs an HTTP client with optional mTLS.
func buildHTTPClient(certEnv, keyEnv, caEnv string, timeout time.Duration) (*http.Client, error) {
	tlsCfg := &tls.Config{
		MinVersion: tls.VersionTLS12,
	}

	if certEnv != "" && keyEnv != "" {
		certPEM, err := readValueOrFile(certEnv)
		if err != nil {
			return nil, fmt.Errorf("failed to read client cert: %w", err)
		}
		keyPEM, err := readValueOrFile(keyEnv)
		if err != nil {
			return nil, fmt.Errorf("failed to read client key: %w", err)
		}
		cert, err := tls.X509KeyPair(certPEM, keyPEM)
		if err != nil {
			return nil, fmt.Errorf("failed to load client certificate/key: %w", err)
		}
		tlsCfg.Certificates = []tls.Certificate{cert}
	} else {
		log.Printf("[signing] mTLS client cert/key not provided; proceeding without client auth")
	}

	if caEnv != "" {
		caPEM, err := readValueOrFile(caEnv)
		if err != nil {
			return nil, fmt.Errorf("failed to read CA cert: %w", err)
		}
		cp := x509.NewCertPool()
		if !cp.AppendCertsFromPEM(caPEM) {
			return nil, errors.New("failed to parse CA certificate")
		}
		tlsCfg.RootCAs = cp
	}

	transport := &http.Transport{
		TLSClientConfig: tlsCfg,
	}

	return &http.Client{
		Transport: transport,
		Timeout:   timeout,
	}, nil
}

// readValueOrFile returns the raw bytes of the provided string. If the string points to an existing
// file path the file contents are returned; otherwise the string itself is treated as PEM content.
func readValueOrFile(value string) ([]byte, error) {
	if value == "" {
		return nil, errors.New("value is empty")
	}
	if _, err := os.Stat(value); err == nil {
		return os.ReadFile(value)
	}
	if strings.Contains(value, "BEGIN") {
		return []byte(value), nil
	}
	// Best-effort base64 decode support for CI-provided secrets.
	if decoded, err := base64.StdEncoding.DecodeString(value); err == nil && len(decoded) > 0 {
		return decoded, nil
	}
	return []byte(value), nil
}

// decodeEd25519Key decodes a base64 private key into an ed25519 keypair.
func decodeEd25519Key(keyB64 string) (ed25519.PrivateKey, ed25519.PublicKey, error) {
	data, err := base64.StdEncoding.DecodeString(strings.TrimSpace(keyB64))
	if err != nil {
		return nil, nil, fmt.Errorf("unable to decode base64 key: %w", err)
	}
	switch len(data) {
	case ed25519.SeedSize:
		priv := ed25519.NewKeyFromSeed(data)
		return priv, priv.Public().(ed25519.PublicKey), nil
	case ed25519.PrivateKeySize:
		priv := ed25519.PrivateKey(data)
		return priv, priv.Public().(ed25519.PublicKey), nil
	default:
		return nil, nil, fmt.Errorf("unexpected ed25519 private key length %d", len(data))
	}
}

func derivePublicKey(keyB64 string) (ed25519.PublicKey, error) {
	_, pub, err := decodeEd25519Key(keyB64)
	return pub, err
}

func shortSHA(data []byte) []byte {
	sum := sha256.Sum256(data)
	return sum[:4] // first 8 hex chars
}

type httpStatusError struct {
	StatusCode int
	Body       string
}

func (e *httpStatusError) Error() string {
	return fmt.Sprintf("kms http %d: %s", e.StatusCode, e.Body)
}

func (e *httpStatusError) ShouldRetry() bool {
	return e.StatusCode >= 500 && e.StatusCode < 600
}
